### ISSUE TITLE
Key Generation Best Practices & Possibly Enforcing Secure Keys

### ISSUE THREAD

**Issue by ankane** [attr-encrypted/attr_encrypted#311]
Hey @saghaulor, what I like most about this gem is makes encryption accessible to anyone with a few lines of code, in a way that's difficult to make mistakes. I think one of the areas that could be improved is key generation. I suspect many developers are generating a random human-readable 32 length string instead of 32 random bytes. It'd be great to give guidance on how to generate a secure key.

For instance, for those who use ENV variables or Rails credentials/secrets for keys, one approach is to create a hex key with `SecureRandom.hex(32)`, then unpack this before passing it to the gem with `[hex_key].pack("H*")`. Once a best practice is established, it'd be nice to have a built-in support for it. For instance, if the user passes a hex key, the gem does the conversion internally.

The [rbnacl gem](https://github.com/crypto-rb/rbnacl) (bindings for libsodium) has a neat way to enforce secure keys. It [checks if the string is binary](https://github.com/crypto-rb/rbnacl/blob/e4d5811b5b75122fade3177a1196f696830309f7/lib/rbnacl/util.rb#L169) before allowing it as a key. This means that `SecureRandom.random_bytes(32)` works as a key and `"This is a key that is 256 bits!!"` does not.

Curious to hear your thoughts. @david-a-wheeler and I have been discussing this here: https://github.com/ankane/blind_index/issues/5

**Comment by softwaregravy** [comment]
Just found this and am very grateful to have found a gem that does exactly what I was looking for. I found this issue while looking for best practices for generating my key. While I'm indifferent to the feature of preventing bad keys, I am very much in favor of providing a best practices section in the main doc. Just adding a section with a few options on best practices to generate they key(s).

**Comment by adiakritos** [comment]
So does the key that I use to encrypt data need to be stored or can it be a unique value every time a variable is stored? Is this what the IV column is for? To decrypt the other value?

**Comment by david-a-wheeler** [comment]
@adiakritos : You must store an encryption key *somewhere* if you want to use it to decrypt data later.  In many environments the encryption key is stored as an environment variable (be *sure* to not display that environment variable to untrusted parties - e.g., don't print it as part of your CI test suite).

The IV column is something different.  Many encryption algorithms need not just the key, but also a random "initialization vector" (IV) that doesn't need to be kept secret but must be stored for later use.  More information here: https://en.wikipedia.org/wiki/Initialization_vector .  For most people, this is just a low-level detail... just create the column & you'll be happy.

**Comment by saghaulor** [comment]
@ankane @david-a-wheeler thanks for the great feedback. I've definitely been tossing around the idea of how to make key handling dead simple. Certainly some best practices in the README will go a long way.

If I could go back to the beginning, knowing what I know today, I'd totally rewrite this gem to serialize everything into a single column, everything would include metadata about the key used to encrypt, so as to make rotation significantly easier. 

I think forcing the key to be binary is a good idea. Given that it would likely break crypto operations for many users, I think we'd have to introduce is as being configurable. We could even have it on by default with upgrade notes indicating that it may cause problems if keys were not generated correctly.

**Comment by saghaulor** [comment]
@ankane at my employer we've already built a solution. I'd have to talk with our legal team to see if I'm allowed to disclose the details, but the gist of it is that the metadata is stored with the encrypted payload, so each object knows how to decrypt itself, no matter what the circumstances of it's encryption were.

The challenge about moving to a solution like this is how do we provide a migration path forward. We've solved that problem as well, but it wasn't pretty.

Additionally, I hesitate to suggest a solution where the key is sitting in your app. Our solution externalized crypto from the Ruby process. I worry that it makes the key less safe.

Regarding the key verification, I wonder if simply validating against binary is sufficient. It seems like unpacking hex to binary may be going to far for the concerns of this gem. A user can already pass a method or proc as a key, this method or proc can contain the necessary logic to unpack the key to binary. 

That being said, I wonder if validating the key material should even be the concern of attr_encrypted. It seems like it should be a concern of the underlying cryto library.
